drivers: video: Add format size

ngphibang opened the pull request:
- Add a size field into video_format struct.
- Video receiver drivers now need to set the format's size to expose to application. Application should base on the format size to allocate buffers. The caps' min/max_line_count fields (which are needed only for HWs that cannot support the whole image frame) can hence be dropped.

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -79,6 +79,15 @@ struct video_format {
+	/**
+	 * @brief size of the buffer in bytes.
+	 *
+	 * For uncompressed formats, this is the size of the raw data buffer in bytes.
```
Any word about fractioned frames?

josuah commented on drivers/video/video_mcux_smartdma.c:
```diff
@@ -271,6 +272,7 @@ static int nxp_video_sdma_get_format(const struct device *dev, struct video_form
 	}
 
 	fmt->pitch = fmt->width * video_bits_per_pixel(fmt->pixelformat) / BITS_PER_BYTE;
+	fmt->size = fmt->pitch * SDMA_LINE_COUNT;
```
This is the minimum size that the driver author can allocate, but has the merit of allowing `fmt->size` to be accurate, and not require any wrapper in `video_set_format()`. Good idea!

josuah reviewed (commented):
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/video_shell.c
This has a print statement for logging the max count

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/video_emul_rx.c
This will be needing some conversion too, useful for testing this change on CI.

This sounds like a better solution than what I proposed here IMHO https://github.com/zephyrproject-rtos/zephyr/pull/92884#discussion_r2413796420

avolmat-st commented:
> > Yes. Cannot be done on the pitch value I think since pitch is related to the way it is organized into memory, which is not applicable for sensors (which do not write themselves into memory).
> 
> For compressed formats, the receiver drivers need to set the pitch to 0 (either it is using the bits_per_pixel() utility or not). I think.

Yes, agree. Compressed format = pitch 0.

> For compressed formats, I also think the sensor should not set the fmt.size. Only the bridge / dma drivers who deals with the memory need to set the pitch / size which are related to memory. The sensor output data in the defined compressed formats (quality ratio, etc.) and the bridge drivers should be able to estimate the size.

Hum, the bridge doesn't have any idea about how the compressed data have been produced. The quality ratio for example is something which is applicable to the sensor itself and not the bridge so the bridge doesn't have much information to estimate the size apart from the format (say, JPEG), width and height, which is very few information and won't help to give a correct estimation of the size. In such case, what would be the estimated JPEG size for example ?

ngphibang commented:
> Another aspect is that size from compressed data is always an estimation, and for a given format, different driver authors could give different estimated sizes.
> 
> For instance, VENC picked an arbitrary width * height / 10, but ultimately, the user knows if they are shooting at tree leaves in a tempest (huge data) or immobile parking (tiny data).
> 
> Would that work around the problem give a per-format expectation but give the real estimation to the user?
> 
> Unless we are after a formula that says "by theory, a frame can never go above this size even for random noise"?

If we let the user to estimate this, it goes back to the original implementation where the application makes the decision.
In fact, the driver has more knowledge and controls on the size than the user. For the dependency on image content (static vs dynamic scene) as you said, the drivers should still think of it and it should give a rough estimate for the worst case.

ngphibang commented:
> Hum, the bridge doesn't have any idea about how the compressed data have been produced. The quality ratio for example is something which is applicable to the sensor itself and not the bridge so the bridge doesn't have much information to estimate the size apart from the format (say, JPEG), width and height, which is very few information and won't help to give a correct estimation of the size. In such case, what would be the estimated JPEG size for example ?
> 
> I agree, that seen from the sensor, there is nothing related to memory usage, however, how to have a "not too bad" estimation ?

We have `VIDEO_CID_JPEG_COMPRESSION_QUALITY` for sensor to help receiver to get better estimate. The problem is the sensor can give information but then the receiver will the decision on choice of size depending on its constraint (DMA constraints, lack of memory like smartdma, etc.). But yes, as you said, I think sensor can set the fmt.size to help a better choice for receiver for compressed formats (anyway, it will be added when there are support for Jpeg on the sensor side)

josuah commented:
> We have VIDEO_CID_JPEG_COMPRESSION_QUALITY for sensor to help receiver to get better estimate

Good point, it is always possible for video receiver to query the information we need.
And if not wanting to put too much smart in the driver it is always possible to have estimatior functions in `video_common.c`. For instance VENC can benefit from that too.

josuah reviewed (dismissed):
There is an open question about how to handle video compression provided by sensors, but both possible options seems like having robust path forward.

LGTM as soon as CI passes for the video shell and CI driver.

Thank you for the quick proposal!

avolmat-st reviewed (dismissed):
Nit: I wrongly set min_vbuf_count / min/max_line_count in mipid02 when I wrote it which is incorrect since the MIPID02 is not a receiver.  It is possible to also remove caps->min_vbuf_count = 1; from its caps.

ngphibang committed:
drivers: video: st_mipid02: Drop caps' min_vbuf_count

The mipid02 is just a bridge driver which does not deal with memory so
should not expose caps' min_vbuf_count.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Set format size

Receiver drivers now need to set the format's size to expose it to
the application.

Application should base on the format size to allocate buffers. The
caps' min/max_line_count (which are needed only for HWs that cannot
support the whole image frame) can hence be dropped.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

avolmat-st reviewed (commented):
Small point regarding the migration-guide

ngphibang committed:
doc: migration-guide-4.3: Mention some changes in video structures

Mention some dropped fields in the video_caps structure and the buffer
allocation size for application.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

avolmat-st reviewed (dismissed):
Thanks @ngphibang. All good to me.

ngphibang committed:
drivers: video: Add helper to estimate format size

Add a helper to estimate format size and pitch.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
doc: release-note-4.3: Add video format estimation helper

Mention a new video_estimate_fmt_size helper.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

avolmat-st reviewed (approved):
LGTM. Thanks a lot.
